### PR TITLE
HPCC-22112 Provide tracing when Record Translation has taken effect

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Observed logs after smoketest run

### CODE REVIEW

**Comment by richardkchapman** [comment]
@jakesmith Please review (especially thor changes)

**Comment by jakesmith** [comment]
A couple of questions:
1) Should DBGLOG's be PROGLOG's ? (and in doDescribe in GeneralRecordTranslator also)
2) Do we need a 'traceTranslations' option in hthor and thor too, or is in roxie only ok?

@richardkchapman

**Comment by richardkchapman** [comment]
I didn't bother making the tracing optional in thor/hthor as I did not expect the volume of tracing to be an issue in those platforms.

As for DBGLOG vs PROGLOG - @shamser can you tell us the current rules?

File: initfiles/componentfiles/configschema/xsd/roxie.xsd [hunk 1]
```diff
@@ -274,6 +274,9 @@
                     <xs:attribute name="soapTraceLevel" type="xs:nonNegativeInteger" hpcc:displayName="Soap trace Level"
                                   hpcc:presetValue="1"
                                   hpcc:tooltip="Level of detail in reporting SOAPCALL information(set to 0 for none, 1 for normal, >1 or more for extended)"/>
+                    <xs:attribute name="traceTranslations" type="xs:boolean" hpcc:displayName="Trace record layout translations"
+                                  hpcc:presetValue="true"
```
Comment (shamser) [inline initfiles/componentfiles/configschema/xsd/roxie.xsd]: Should this tracing be enabled by default?  As it's debug tracing, shouldn't this be disabled by default?

**Review by shamser — COMMENTED** [review]
@richardkchapman

File: initfiles/componentfiles/configschema/xsd/roxie.xsd (hunk 1 above)
Comment (richardkchapman) [reply initfiles/componentfiles/configschema/xsd/roxie.xsd]: It's tracing the users want to look out for... Does that mean it should be a PROGLOG?

I've never been particularly clear on the rules for using the various logging macros.

**Comment by richardkchapman** [comment]
@ghalliday Please review/merge

**Comment by jakesmith** [comment]
I thought the idea was that at some point we'd disabled DBGLOG in release builds.
So if the tracing is still wanted in release, we should use something else (like PROGLOG).

File: initfiles/componentfiles/configschema/xsd/roxie.xsd (hunk 1 above)
Comment (shamser) [reply initfiles/componentfiles/configschema/xsd/roxie.xsd]: DBGLOG is temporary logging used for debugging/investigation purposes (ithe plan is to have DBGLOG disabled by default.).  For user information, it should be LOG(MCuserInfo, format,...).  There is an explaination of the difference types near the top of jlog.hpp.

**Comment by ghalliday** [comment]
@richardkchapman  is this suitable for a 7.4 release?
The tracing could get very large - e.g. some nasty dataset with nested complex datasets.  Should there be a compact description of what is being translated?

**Comment by richardkchapman** [comment]
@ghalliday I have tweaked the tracing a bit to bake it more compact and also more useful.

I don't want to change the DBGLOGs at this time:

1. None of the message categories listed in jlog seem more appropriate. It's not "progress", nor is it a warning or error. It's information, and DBGLOG is closest to covering that. 
2. There are literally 1000's of uses of DBGLOG throughout the code using the same assumption.
3. If we need to look at all DBGLOG and PROGLOG to determine which are progress, which are info, and which are "temporary to be removed in release mode" then that's a separate task, but I'm not sure the last of those is a useful concept (we can, and do, put DBGLOG's inside of #ifdef's)

**Comment by ghalliday** [comment]
@richardkchapman first scan looks good.  Can you squash and I'll recheck it.

**Comment by richardkchapman** [comment]
@ghalliday Squashed

**Comment by ghalliday** [comment]
Slight concern about the size of the output, but we can monitor.

**Pull request merged**